File: cli/src/main/java/io/bitken/ss/cli/store/StateReport.java
package io.bitken.ss.cli.store;

import io.bitken.ss.cli.ResolutionJson;
import io.bitken.ss.cli.conf.ds.ProjectDataStore;
import io.bitken.ss.conf.ResolvedStateRoot;
import io.bitken.ss.conf.ShipsmoothDataLocator;

import java.nio.file.Path;

/**
 * Shared "where does state live" reporting for the {@code store} commands ({@code info} and
 * {@code init}'s success output). Emits the {@code ready} shape — {@code storageType}, the
 * state root, and the ready-to-read {@code plansDir} — as either a machine-readable JSON line
 * or human text. All output goes to stdout.
 */
final class StateReport {

    private StateReport() {
    }

    /** Print the ready/settled state report for {@code store} resolved at {@code repoRoot}. */
    static void printReady(Path repoRoot, ProjectDataStore store, boolean json) {
        Path stateRoot = store.stateRoot();
        String storageType = store instanceof ProjectDataStore.InRepo ? "same-repo" : "separate-dir";
        // plansDir via the locator so the same-repo (.shipsmooth) vs separate-dir layout difference
        // stays owned by the single source of path truth, not re-derived here.
        Path plansDir = new ShipsmoothDataLocator(repoRoot, ResolvedStateRoot.of(stateRoot)).plansDir();

        if (json) {
            System.out.println(ResolutionJson.ready(storageType, stateRoot, plansDir));
        } else {
            System.out.println("shipsmooth: " + storageType + " storage at " + stateRoot);
            System.out.println("plans: " + plansDir);
        }
    }
}

File: exp/rust/crates/ss-cli/src/store/state_report.rs
//! Shared "where does state live" reporting for the `store` commands
//! (`info` and, with Task 7, `init`'s success output).
//!
//! Port of the Java `StateReport`. Emits the `ready` shape — `storageType`,
//! the state root, and the ready-to-read `plansDir` — as either a
//! machine-readable JSON line or human text.

use std::path::Path;

use ss_core::conf::{ResolvedStateRoot, ShipsmoothDataLocator};

use crate::ds::store::ProjectDataStore;
use crate::resolution_json;

/// The ready/settled state report for a store resolved at `repo_root`.
pub fn ready(repo_root: &Path, store: &ProjectDataStore, json: bool) -> String {
    let state_root = store.state_root();
    let storage_type = match store {
        ProjectDataStore::InRepo { .. } => "same-repo",
        ProjectDataStore::Standalone { .. } => "separate-dir",
    };
    // plansDir via the locator so the same-repo (.shipsmooth) vs separate-dir
    // layout difference stays owned by the single source of path truth, not
    // re-derived here. A settled store's roots were just verified by the
    // resolver, so token minting cannot fail short of a filesystem race.
    let token = ResolvedStateRoot::of(state_root).expect("settled state root must be accessible");
    let plans_dir = ShipsmoothDataLocator::new(repo_root, token)
        .expect("settled repo root must be accessible")
        .plans_dir();

    if json {
        resolution_json::ready(storage_type, state_root, &plans_dir)
    } else {
        format!("shipsmooth: {storage_type} storage at {}\nplans: {}", state_root.display(), plans_dir.display())
    }
}
